presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java

```diff
@@ -138,6 +141,9 @@ public PrestoSparkTaskExecution(
                 "Fragment is partitioned, but not all partitioned drivers were found");
 
         taskHandle = createTaskHandle(taskStateMachine, taskContext, localExecutionPlan, taskExecutor);
+
+        requireNonNull(memoryUpdateExecutor, "memoryUpdateExecutorService is null");
+        memoryUpdateExecutor.schedule(taskContext::updatePeakMemory, 5, SECONDS);
```

Comment:
We can also make `taskContext` itself to refresh the peak memory usage periodically. What do you think? @arhimondr , @tdcmeehan ?

Reply:
How does it work in Presto classic? How is it getting refreshed?

Reply:
From what I can tell, there's many ways it could be updated, but primarily it's done by `TaskInfoFetcher` calling into the tasks, which will call:

`TaskResource#getTaskInfo` -> `SqlTaskManager#getTaskInfo` -> `SqlTask#getTaskInfo` -> `SqlTask#createTaskInfo` -> `SqlTask#getTaskStats` -> `TaskContext#getTaskStats`

The default interval there is `1s`, but it could be quicker if the `TaskStatus` changes state before then (e.g. finishes).  I'm guessing here, there is nothing polling the info/status, hence the need to schedule this update from within the task?  

Also, would it be inaccurate if the status transitions before the `memoryUpdateExecutor` executes its next update?  For example upon failure.

Reply:
I think the idea before was the stats were refreshed whenever there is a heartbeat from coordinator to worker/task.  Is there such a heartbeat in PoS we can piggy back off of?

Reply:
@arhimondr : 

> How does it work in Presto classic? How is it getting refreshed?

The peak memory stats in TaskContext only get updated when
TaskContext::getTaskStats(). Presto-on-Spark doesn't pull task stats
actively so it needs to be updated explicitly.

Reply:
@tdcmeehan  : 

> I'm guessing here, there is nothing polling the info/status, hence the need to schedule this update from within the task?

Exactly.

Reply:
If that's the case, could we add a listener to `taskStateMachine`, and just update the stats when the state changes (as I presume nothing is looking at the stats in-between state transitions?)

Reply:
@tdcmeehan : But peak memory only goes up and down when state is `RUNNING` right?

Reply:
It feels like it is reasonable to add an explicit `updatePeakMemory` method to the `TaskContext` and call it periodically. Given it is not needed for classic Presto i would keep the responsibility of periodically calling it on Presto on Spark code.

Reply:
@arhimondr : Agreed. That's why i did it in this way. 

Although I think make `TaskContext` do periodically update peak memory is not a bad idea (today in Presto classic it relies on periodic "pull" to get updated)... but i will leave as future disucssion/work